### PR TITLE
net: http_client: receiver retry on EAGAIN

### PR DESCRIPTION
**Pull request by oopsmonk**
`http_client_req` is unable to get response data from socket because data is not ready when `http_wait_data` calling `revc`.
it happened on b_l4s5i_iot01a target but not stm32f746g_disco.

```
[00:00:14.924,000] <dbg> net_http.Data to send
                                  47 45 54 20 2f 67 65 74  20 48 54 54 50 2f 31 2e |GET /get  HTTP/1.
                                  31 0d 0a 48 6f 73 74 3a  20 68 74 74 70 62 69 6e |1..Host:  httpbin
                                  2e 6f 72 67 3a 34 34 33  0d 0a 43 6f 6e 74 65 6e |.org:443 ..Conten
                                  74 2d 54 79 70 65 3a 20  61 70 70 6c 69 63 61 74 |t-Type:  applicat
                                  69 6f 6e 2f 6a 73 6f 6e  0d 0a 0d 0a             |ion/json ....    
[00:00:14.937,000] <dbg> wifi_eswifi.eswifi_spi_request: cmd=0x20000e5d (5 byte), rsp=0x20000e5d (1600 byte)
[00:00:14.954,000] <dbg> wifi_eswifi.eswifi_spi_request: success
[00:00:15.104,000] <dbg> net_http.http_client_req: (shell_uart): Sent 135 bytes
[00:00:15.112,000] <dbg> wifi_eswifi.eswifi_socket_recv: read -11 0 (nil)
[00:00:15.119,000] <dbg> net_http.http_wait_data: Connection error (11)
[00:00:15.126,000] <dbg> net_http.http_client_req: (shell_uart): Wait data failure (-11)
[00:00:15.135,000] <dbg> iota_http.http_client_get: data: 
[00:00:15.158,000] <dbg> wifi_eswifi.eswifi_off_read_work: 
[00:00:15.187,000] <dbg> wifi_eswifi.__stop_socket: Stopping socket 0
[00:00:15.194,000] <dbg> wifi_eswifi.eswifi_spi_request: cmd=0x200035f8 (5 byte), rsp=0x20000e5d (1600 byte)
[00:00:15.215,000] <dbg> wifi_eswifi.eswifi_spi_request: success
```

with reties, the response can be processed by http_client

```
[00:00:15.343,000] <dbg> net_http.Data to send
                                  47 45 54 20 2f 67 65 74  20 48 54 54 50 2f 31 2e |GET /get  HTTP/1.
                                  31 0d 0a 48 6f 73 74 3a  20 68 74 74 70 62 69 6e |1..Host:  httpbin
                                  2e 6f 72 67 3a 34 34 33  0d 0a 43 6f 6e 74 65 6e |.org:443 ..Conten
                                  74 2d 54 79 70 65 3a 20  61 70 70 6c 69 63 61 74 |t-Type:  applicat
                                  69 6f 6e 2f 6a 73 6f 6e  0d 0a 0d 0a             |ion/json ....    
[00:00:15.501,000] <dbg> wifi_eswifi.eswifi_spi_request: cmd=0x20000e5d (83 byte), rsp=0x20000e5d (1600 byte)
[00:00:15.519,000] <dbg> wifi_eswifi.eswifi_spi_request: success
[00:00:15.525,000] <dbg> net_http.http_client_req: (shell_uart): Sent 135 bytes
[00:00:15.533,000] <dbg> wifi_eswifi.eswifi_socket_recv: read -11 0 (nil)
[00:00:15.540,000] <dbg> net_http.http_wait_data: receiver recv retry (0)
[00:00:15.578,000] <dbg> wifi_eswifi.eswifi_off_read_work: 
[00:00:15.648,000] <dbg> wifi_eswifi.eswifi_spi_request: cmd=0x20003cac (3 byte), rsp=0x20000e5d (1600 byte)
[00:00:15.694,000] <dbg> wifi_eswifi.eswifi_spi_request: success
[00:00:15.700,000] <dbg> wifi_eswifi.eswifi_socket_recv: read -11 0 (nil)
[00:00:15.707,000] <dbg> net_http.http_wait_data: receiver recv retry (1)
[00:00:15.800,000] <dbg> wifi_eswifi.eswifi_off_read_work: 
[00:00:15.875,000] <dbg> wifi_eswifi.eswifi_spi_request: cmd=0x20003cac (3 byte), rsp=0x20000e5d (1600 byte)
[00:00:15.912,000] <dbg> wifi_eswifi.eswifi_spi_request: success
[00:00:15.918,000] <dbg> wifi_eswifi.eswifi_off_read_work: payload sz = 472
[00:00:15.925,000] <dbg> wifi_eswifi.eswifi_off_read_work: 
[00:00:15.931,000] <dbg> wifi_eswifi.eswifi_socket_recv: read 472 0 0x200085e8
[00:00:15.939,000] <dbg> wifi_eswifi.eswifi_spi_request: cmd=0x20000e5d (5 byte), rsp=0x20000e5d (1600 byte)
[00:00:15.949,000] <dbg> net_http.on_message_begin: (shell_uart): -- HTTP GET response (headers) --
[00:00:15.953,000] <dbg> net_http.on_status: (shell_uart): HTTP response status 200 OK
```

This PR added a retry mechanism `http_wait_data` and configurations in Kconfig for handling the EAGAIN error.

```
[*] HTTP client API [EXPERIMENTAL]
(5)     HTTP client receiver retry
(100)   HTTP client retry interval in ms
```

### CODE REVIEW

File: subsys/net/lib/http/http_client.c [hunk 1]
```diff
@@ -423,6 +424,17 @@ static int http_wait_data(int sock, struct http_request *req)
...
 			break;
+		} else if (received == -EAGAIN) {
+			/* Retry */
+			if (retry > CONFIG_HTTP_CLIENT_RECV_RETRY) {
+				LOG_ERR("Receiver retry failed (%d)", errno);
+				ret = -errno;
+				break;
+			}
+			LOG_DBG("Receiver retry (%d)", retry);
+			k_msleep(CONFIG_HTTP_CLIENT_RETRY_INTERVAL_MS);
```
Comment (jukkar) [inline subsys/net/lib/http/http_client.c:435]: This will block the application now and waits always the timeout. Could we use `poll()` so that if we receive the data, we can immediately call `recv()` instead of waiting possibly long time.

**Review by pfalcon — CHANGES_REQUESTED** [review]
This appears to be misplaced workaround for a bug in particular driver. The rules for producing EAGAIN from send()/recv() operations:

* In default socket configuration (blocking), EAGAIN may never be returned.
* With non-blocking socket, EAGAIN means that data can't be sent/received currently. Blindly retrying operation doesn't make sense, it will result in the same error and 100% CPU burning for extended time. On EAGAIN, poll() should be called instead for non-blocking sockets.
* If SO_RCVTIMEO/SO_SNDTIMEO socket options enabled, EAGAIN means timeout occurred. Again, it should not be retried - user specifically set it to deal with it on application side.

So, please look into fixing the driver to work per the spec above instead.

**Comment by oopsmonk** [comment]
> This appears to be misplaced workaround for a bug in particular driver. The rules for producing EAGAIN from send()/recv() operations:
> 
>     * In default socket configuration (blocking), EAGAIN may never be returned.
> 
>     * With non-blocking socket, EAGAIN means that data can't be sent/received currently. Blindly retrying operation doesn't make sense, it will result in the same error and 100% CPU burning for extended time. On EAGAIN, poll() should be called instead for non-blocking sockets.
> 
>     * If SO_RCVTIMEO/SO_SNDTIMEO socket options enabled, EAGAIN means timeout occurred. Again, it should not be retried - user specifically set it to deal with it on application side.
> 
> 
> So, please look into fixing the driver to work per the spec above instead.

Thanks for the information, I'll try to look into eswifi driver.

**Review by rlubos — CHANGES_REQUESTED** [review]
I agree with @pfalcon, the library uses blocking socket so there should be no EAGAIN, please fix the respective driver instead.

**Comment by jukkar** [comment]
>  the library uses blocking socket so there should be no EAGAIN, please fix the respective driver instead

Hmm, the http-client library uses whatever socket is given to it, it can be blocking or non-blocking. There is a timeout option in `http_client_req()` that gives user an option to specify how long to wait for the data to arrive.

Note sure about the issue that this PR is trying to fix, it indeed looks like the driver is returning wrong status.

**Comment by oopsmonk** [comment]
http-client library has the socket timeout option but there is no timeout in this case. 
`recv` is non-blocking in eswifi driver implementation. it might return EAGAIN in purpose?

once the socket connection is created, `eswifi_off_read_work` starts polling data every 100ms, when http-client calls `http_wait_data` data is not ready because socket fifo is empty and `recv` is non-blocking. the result is http-client gets an error and close the socket before data arriving in fifo.

I agree this PR might be an ugly workaround because the driver doesn't support blocking mode, or did I miss something? 
In the normal case, retry won't happen.

**Comment by rlubos** [comment]
> recv is non-blocking in eswifi driver implementation. it might return EAGAIN in purpose?

If your application did not put the socket into non-blocking mode, then it doesn't sound good, the socket should be blocking by default when created. It might be a nasty surprise not only in this case, but also in others.

> Hmm, the http-client library uses whatever socket is given to it, it can be blocking or non-blocking. 

Sorry, I missed the fact that it's not the library that creates the socket, was merely looking for code that would put the socket into non-blocking mode. Nonetheless, the `http_wait_data()` implementation clearly suggests it expects a blocking socket to work with. So if the library should not make any assumptions on the underlying socket mode, switching to poll/recv combo sounds fine as well.

File: subsys/net/lib/http/http_client.c [hunk 2]
```diff
@@ -408,49 +409,63 @@ static void http_client_init_parser(struct http_parser *parser,
...
+
+	ret = poll(&poll_fd, 1, timeout);
+	if (ret < 0) {
+		LOG_ERR("Socket poll err (%d)", errno);
+		return -errno;
+	} else if (ret == 0) {
+		LOG_ERR("Socket poll timeout");
+		return ret;
+	} else if (poll_fd.revents & POLLIN) {
+		do {
```
Comment (rlubos) [inline subsys/net/lib/http/http_client.c:430]: The internal do/while loop doesn't seem right - `recv()` might've read all of the data available,  but still expect more, so it'll block anyway. A proper way would be to `poll()` at the beginning of each loop iteration. It'd also keep the diff smaller - no extra indentation.

Comment (oopsmonk) [reply subsys/net/lib/http/http_client.c:430]: ah, you are right. if I read it correctly, you mean
```c
do{
  poll();
  if(ret<0){
    ret =  -errno;
    break;
  }else if(ret == 0){
    break;
  }
  recv();
  ...
  // break if message_complete or errno
} while(true)
```
right?

Comment (rlubos) [reply subsys/net/lib/http/http_client.c:430]: Yeah, something like this. It'd also make sense to check if `POLLIN` was returned, because if not, it's likely an error condition (`POLLERR` etc).

**Comment by oopsmonk** [comment]
> If your application did not put the socket into non-blocking mode, then it doesn't sound good, the socket should be blocking by default when created. It might be a nasty surprise not only in this case, but also in others.
> 
the socket flag is `SOCK_STREAM` instead of `SOCK_STREAM | O_NONBLOCK` so `recv` should be in blocking mode.

**Review by rlubos — APPROVED** [review]

File: subsys/net/lib/http/Kconfig [hunk 3]
```diff
@@ -29,20 +29,6 @@ config HTTP_CLIENT
...
-	help
-	  The number of retries when EAGAIN happens
-
-config HTTP_CLIENT_RETRY_INTERVAL_MS
-	int "HTTP client retry interval in ms"
-	depends on HTTP_CLIENT
-	default 100
-	help
-	  HTTP client retry interval in ms
-
```
Comment (jukkar) [inline subsys/net/lib/http/Kconfig]: You can squash this change to previous commit if the kconfig options are not used after all. No point introducing a change in one commit and then remove it in next.

File: subsys/net/lib/http/http_client.c [hunk 4]
```diff
@@ -149,6 +149,7 @@ static int on_url(struct http_parser *parser, const char *at, size_t length)
 	struct http_request *req = CONTAINER_OF(parser,
 						struct http_request,
 						internal.parser);
+
```
Comment (jukkar) [inline subsys/net/lib/http/http_client.c:152]: This seems to be unrelated change not related to this commit. Please remove it.

File: subsys/net/lib/http/http_client.c [hunk 5]
```diff
@@ -208,7 +209,7 @@ static int on_header_field(struct http_parser *parser, const char *at,
 	return 0;
 }
 
-#define MAX_NUM_DIGITS	16
+#define MAX_NUM_DIGITS  16
```
Comment (jukkar) [inline subsys/net/lib/http/http_client.c:212]: Same comment to this one. If you want to fix these white space issues, then you can introduce a separate commit that does that (could be part of this PR of course)

**Review by jukkar — COMMENTED** [review]
Looks good, just minor comment

File: subsys/net/lib/http/Kconfig (hunk 3 above)
Comment (oopsmonk) [reply subsys/net/lib/http/Kconfig]: sure, I thought PRs are merged via GitHub's squash and merge, but this project prefers a local squash.

File: subsys/net/lib/http/http_client.c (hunk 4 above)
Comment (oopsmonk) [reply subsys/net/lib/http/http_client.c:152]: it is modified by uncrustify, I will move it to a new commit, maybe remove it if uncrustify has no complaints. :p 
https://docs.zephyrproject.org/latest/contribute/index.html#uncrustify

File: subsys/net/lib/http/http_client.c (hunk 5 above)
Comment (oopsmonk) [reply subsys/net/lib/http/http_client.c:212]: ditto

File: subsys/net/lib/http/http_client.c (hunk 4 above)
Comment (oopsmonk) [reply subsys/net/lib/http/http_client.c:152]: introduced 51dbeb6 to fix coding style issues, the new line is needed.

**Review by jukkar — APPROVED** [review]

**Comment by oopsmonk** [comment]
@pfalcon any suggestions on current changes?

**Comment by pfalcon** [comment]
@jukkar commented 8 days ago

>>    the library uses blocking socket so there should be no EAGAIN, please fix the respective driver instead

> Hmm, the http-client library uses whatever socket is given to it, it can be blocking or non-blocking.

Working with blocking vs non-blocking sockets require radically different approaches (*). Working with blocking sockets is much simpler. A library written to work with blocking sockets (== a library written to work in a simple way) will NOT work with non-blocking sockets. A reason to use non-blocking sockets is to benefit from light-weight I/O concurrency, i.e. being able to schedule multiple operations at the same time (without heavy-weight threads).

Now you can tell whether `subsys/net/lib/http` adheres to any of the above criteria. (From the lack of poll() calls in it, or integration with generic event loop, it's easy to say it doesn't support non-blocking sockets.)

(*) That is, to fully benefit from non-blocking sockets. One of course can use non-blocking sockets as of they were blocking, but ... then one should just use the blocking sockets directly.

> There is a timeout option in http_client_req() that gives user an option to specify how long to wait for the data to arrive.

I checked it now, and it's implemented in a pretty round-about way - there's a delayed work, which *closes* socket to get out of recv() call (I assume it does that (it should), and we a test for that behavior). It's a bit strange, because it effectively closes socket behind user's back in this case. It's also not clear how this (timeout) condition is signalled to the user - http_client_req() seems to return `total_sent` in this case nonetheless, nor I see anything special (i.e. error indicator) being passed to a user callback.

All in all, it would be better to rework this timeout implementation, now that we have native socket timeouts implemented.


> Not sure about the issue that this PR is trying to fix, it indeed looks like the driver is returning wrong status.

Ack.

**Review by pfalcon — CHANGES_REQUESTED** [review]
> @pfalcon any suggestions on current changes?

Yes, sadly I have to reinstate my -1. I very rarely give -1 nowadays, but here I think the change here is mis-direct and of no value.

First of all, you report that the issue you face happens with "b_l4s5i_iot01a", and yourself say that it doesn't happen with another board. You were suggested that that looks like a network driver for that board (which driver uses network/socket offloading, and such drivers ofetntimes have problems with correct implementation of socket semantics). You seem to have explored that and confirmed that eswifi doesn't work as expected, but instead of fixing that driver, you again proceeded with adding workaround for particular hardware to a hardware-independent library.

That's not the right approach. The change you make in the latest patch is not needed for blocking sockets (with which http_client library is intended to work). And it's not sufficient to make it work with non-blocking sockets (because with such, both send() and recv() operations are non-blocking). Besides, it makes no sense to "support" non-blocking sockets with this library, as it would need to be completely redesigned and rewritten to get any benefit from non-blocking operations.

So, I would like to extend my suggestion to investigate eswifi driver issues, and fix them on its side. This way, boards using it will work with any socket networking library as expected, be that HTTP, MQTT, LWM2M, or something else. (And again, we can't patch them all to workaround broken drivers.)

**Comment by oopsmonk** [comment]
@pfalcon I agree, appreciate your suggestions. 
The problem is caused by a return error from eswifi driver, this change might fix the problem in `subsys/net/lib/http` but it still caused problems on other network libs. I'm unable to fix the driver issue at this moment, will close this PR and create a new issue. thanks.

**Comment by pfalcon** [comment]
> The problem is caused by a return error from eswifi driver,

Yes, if eswifi's hardware recv() always works as if it was non-blocking, then you would need to fix it by adding wait-for-event/retry (i.e. analog of poll()) on *driver's* side (and only if socket is in blocking mode). Feel free to give that a try when you got a chance, I'm sure that other developers would help to get it right. Thanks.
